services/api-server/tests/unit/test_services_rabbitmq.py

```diff
@@ -199,14 +199,107 @@ async def test_multiple_producers_and_single_consumer(
 
     # check it received
     assert consumer_message_handler.await_count == 1
-    (data,) = consumer_message_handler.call_args[0]
-    assert isinstance(data, bytes)
-    received_message = LoggerRabbitMessage.parse_raw(data)
+    (job_log,) = consumer_message_handler.call_args[0]
+    assert isinstance(job_log, JobLog)
+
+    assert job_log.job_id == project_id
+    assert job_log.node_id == node_id
+    assert job_log.messages == ["expected message"] * 3
+
+
+#
+# --------------------
+#
+
+
+async def test_one_job_multiple_registrations(
+    log_distributor: LogDistributor, project_id: ProjectID
+):
+    async def _(job_log: JobLog):
+        pass
+
+    await log_distributor.register(project_id, _)
+    with pytest.raises(LogStreamerRegistionConflict) as e_info:
+        await log_distributor.register(project_id, _)
+    await log_distributor.deregister(project_id)
+
+
+async def test_log_distributor_register_deregister(
+    project_id: ProjectID,
+    node_id: NodeID,
+    log_distributor: LogDistributor,
+    produce_logs: Callable,
+    faker: Faker,
+):
+    collected_logs: list[str] = []
+
+    async def callback(job_log: JobLog):
+        for msg in job_log.messages:
+            collected_logs.append(msg)
+
+    published_logs: list[str] = []
 
-    assert received_message.user_id == user_id
-    assert received_message.project_id == project_id
-    assert received_message.node_id == node_id
-    assert received_message.messages == ["expected message"] * 3
+    async def _log_publisher():
+        for _ in range(5):
+            msg: str = faker.text()
+            await asyncio.sleep(0.1)
+            await produce_logs("expected", project_id, node_id, [msg], logging.DEBUG)
+            published_logs.append(msg)
+
+    await log_distributor.register(project_id, callback)
+    publisher_task = asyncio.create_task(_log_publisher())
+    await asyncio.sleep(0.1)
+    await log_distributor.deregister(project_id)
+    await asyncio.sleep(0.1)
+    await log_distributor.register(project_id, callback)
+    await asyncio.gather(publisher_task)
+    await asyncio.sleep(0.5)
+    await log_distributor.deregister(project_id)
+
+    assert len(log_distributor._log_streamers.keys()) == 0
+    assert len(collected_logs) > 0
+    assert set(collected_logs).issubset(
+        set(published_logs)
+    )  # some logs might get lost while being deregistered
```

Comment:
I see what you have done here. Can this be written in a more predictable way. I'm not a fan of these sleeps, they usually tend to cause flaky tests.
You could for example invoke the `_log_publisher()` after subscribing and unsubscirbing and ontain the same result.

Reply:
Yes, I agree. I have another test which does that. I added this test here to make sure that a user can subscribe, unsubscribe and then subscribe again as there was a bug in the previous version when doing exactly this. So here it is on purpose, which why I also do the "weak" check of only receiving a subset of the published logs.